python/katana/local/_graph.pyx

```diff
@@ -144,6 +146,12 @@ cdef class GraphBase:
         """
         return iter(range(self.num_nodes()))
 
+    def edge_iterator(self):
+        """
+        Returns an iterator over the edges of this graph.
+        """
+        return EdgeIterator(self)
```

Comment:
Everywhere else in this API we use IDs instead of the pairs. I think you are right that we should move to a more OO approach to this where we provide more data about the edges. However, I don't think this is the way to do it. The problem is now there are two different ways to access the edges.

I would much prefer having an Edge class which is the pair you are using in this iterator, and also represents the ID of the edge. That way there is a single cohesive way to talk about edges. We have been talking about moving to a dataframe oriented API and I think this is a great chance. We should make the `edges` method return a dataframe like object holding the edges as triples of ID, src, dst and also providing transparent access to the properties on those edges.

Reply:
The fact that we never use pairs and only use IDs is one of the biggest hurdles in converting to PyG. The PyG graph representation is just an mx2 tensor holding the edge list of the graph (basically COO). I don't hate the idea of an edge object instead, so long as its easy to get both endpoints of that edge.

I also like the idea of using pandas DataFrames to hold the edges since this works nicely with lots of python libraries.